### PR TITLE
Add Ostream log exporter and tests

### PR DESCRIPTION
**Pull request by xukaren**
This PR adds:
* A OStream Log Exporter 
* Unit tests and a full logging pipeline test  

cc @MarkSeufert @alolita

### CODE REVIEW

File: exporters/ostream/include/opentelemetry/exporters/ostream/log_exporter.h
Comment (g-easy) [inline exporters/ostream/include/opentelemetry/exporters/ostream/log_exporter.h:12]: I think we should avoid defining namespaces like this in a header.

File: exporters/ostream/src/log_exporter.cc
Comment (g-easy) [inline exporters/ostream/src/log_exporter.cc:43]: What guarantees that record->severity won't be out of bounds?

Comment (MarkSeufert) [reply exporters/ostream/src/log_exporter.cc:43]: I'm thinking that the only way this would be out of bounds is if the user manually sets the severity to a value larger than 24, which would be difficult for them to do since a [class enum](https://github.com/open-telemetry/opentelemetry-cpp/blob/master/api/include/opentelemetry/logs/log_record.h#L39) doesn't let the user implicitly cast from int.

File: api/include/opentelemetry/logs/logger.h
Comment (ThomsonTan) [inline api/include/opentelemetry/logs/logger.h:56]: I think we are going to rename the method name from `log()` to `Log()`. So make sure the new comment matches the new name.

Comment (xukaren) [reply api/include/opentelemetry/logs/logger.h:56]: Once #422 is merged, it should ideally overwrite all the changes in this PR from `logger.h` and `log_record.h`. I included a subset of the changes from the other PR here so the CI would run properly. Currently since the log() method is named with a lowercase, I think it would be good to keep the comment consistent here.

Comment (xukaren) [reply api/include/opentelemetry/logs/logger.h:56]: Currently since the log() method is named with a lowercase, I think it would be good to keep the comment consistent with the code. This change could be part of #422?

File: exporters/ostream/test/ostream_log_test.cc
Comment (MarkSeufert) [inline exporters/ostream/test/ostream_log_test.cc:9]: Remove the duplicated includes?

File: exporters/ostream/test/ostream_log_test.cc
Comment (MarkSeufert) [inline exporters/ostream/test/ostream_log_test.cc:106]: nit: call the exporter's Export() directly to avoid having to create a processor instance?

Comment (xukaren) [reply exporters/ostream/test/ostream_log_test.cc:106]: Thanks, great suggestion! I'll leave testing with a processor for the Integration Test below

**Comment by xukaren** [comment]
@lalitb @g-easy @ThomsonTan Would love another review on this!

File: exporters/ostream/src/log_exporter.cc
Comment (xukaren) [inline exporters/ostream/src/log_exporter.cc:31]: question: wondering if here it would make sense for `MakeRecordable()` should return a `nullptr` if the exporter is already shutdown?

Comment (lalitb) [reply exporters/ostream/src/log_exporter.cc:31]: No strong preference, though it will allow us to fail early.

Comment (xukaren) [reply exporters/ostream/src/log_exporter.cc:31]: Will make the change then if there's no other reason against it :)

**Review by lalitb — APPROVED** [review]
LGTM

File: exporters/ostream/src/log_exporter.cc
Comment (xukaren) [reply exporters/ostream/src/log_exporter.cc:31]: Hmm on second thought, might be better to leave it as is actually, in case the caller is unaware of the reason `nullptr` is being thrown (e.g. exception or shutdown) -  might be later to be introduced consistently across traces/metrics/logs later - so I've left it unchanged :)

File: exporters/ostream/include/opentelemetry/exporters/ostream/log_exporter.h
Comment (ThomsonTan) [inline exporters/ostream/include/opentelemetry/exporters/ostream/log_exporter.h:106]: Do we need output `{}` to wrap the whole map, like `{{key: value}, {key1: value1}}`, or just as `{key: value, key1: value1}`?

File: exporters/ostream/src/log_exporter.cc
Comment (ThomsonTan) [inline exporters/ostream/src/log_exporter.cc:52]: Does this mean log an SDK internal error here? Perhaps add a TODO if so.

File: exporters/ostream/include/opentelemetry/exporters/ostream/log_exporter.h
Comment (xukaren) [reply exporters/ostream/include/opentelemetry/exporters/ostream/log_exporter.h:106]: the outer `{}` around the map is done in the main `Export()` function [here](https://github.com/open-telemetry/opentelemetry-cpp/pull/430/files#diff-5bab08f8a94686dae0568791dbdee4c097ead2922ae0c855436326dc52ccf145R79)

Comment (xukaren) [reply exporters/ostream/include/opentelemetry/exporters/ostream/log_exporter.h:106]: This is the [expected output](https://github.com/open-telemetry/opentelemetry-cpp/pull/430/files#diff-30ea5c00a60dd17b4b14adefe2b4bee8d41888ef8afc74703862e9b3514d19f6R220-R221) I imagined

File: exporters/ostream/src/log_exporter.cc
Comment (xukaren) [reply exporters/ostream/src/log_exporter.cc:52]: Yes, same as #408. Okay wrap with a `TODO`

File: exporters/ostream/include/opentelemetry/exporters/ostream/log_exporter.h
Comment (ThomsonTan) [reply exporters/ostream/include/opentelemetry/exporters/ostream/log_exporter.h:106]: Ok, the caller outputs `{}` and assumed its content is map. Better to output `{}` in `printMap` in case it is called somewhere else.

Comment (xukaren) [reply exporters/ostream/include/opentelemetry/exporters/ostream/log_exporter.h:106]: Good call, will make that change

File: exporters/ostream/src/log_exporter.cc
Comment (ThomsonTan) [inline exporters/ostream/src/log_exporter.cc:61]: Perhaps allocates 1 more char for tailing `\0`, so no need to conver to `std::string` for output.

File: exporters/ostream/src/log_exporter.cc
Comment (ThomsonTan) [inline exporters/ostream/src/log_exporter.cc:77]: nit: `<< std::endl` instead of `<< "\n"`.

Comment (xukaren) [reply exporters/ostream/src/log_exporter.cc:77]: Related- should traces' [span exporter](https://github.com/open-telemetry/opentelemetry-cpp/blob/master/exporters/ostream/src/span_exporter.cc#L64) be updated as well? Or should we leave it here as `\n` to keept it consistent?

**Review by ThomsonTan — APPROVED** [review]

File: exporters/ostream/src/log_exporter.cc
Comment (ThomsonTan) [reply exporters/ostream/src/log_exporter.cc:77]: Thanks for pointing it out, I think we could leave this to keep it consistent.

File: exporters/ostream/src/log_exporter.cc
Comment (xukaren) [reply exporters/ostream/src/log_exporter.cc:61]: Haha sorry again, but maybe this optimization is also needed for [traces](https://github.com/open-telemetry/opentelemetry-cpp/blob/master/exporters/ostream/src/span_exporter.cc#L65)?

File: exporters/ostream/include/opentelemetry/exporters/ostream/log_exporter.h
Comment (jsuereth) [inline exporters/ostream/include/opentelemetry/exporters/ostream/log_exporter.h:53]: if "print_value" and "OwndedValueVisitor" are helper functions, why not just put them into the `.cc` file?

Yes, you have to additionally pass sout_, but you'd still get your template-value-static-do-the-right-thing-with-my-type behavior you want, and we won't have a bunch of weak symbols to resolve later when linking *or* public methods that shouldn't be public.


Something like:

```
template <typename T>
void print_value(const T& item, std::ostream& sout) { ... }
template <typename T>
void print_value(const std::vector<T>& item, std::ostream& sout) { ...}
template <typename T>
void print_value(sdk::common::OwnedAttributeVisitor& item, std::ostream& sout) { ...}
```

**Review by jsuereth — CHANGES_REQUESTED** [review]
Looks good, would just like to move helper methods into CC file or (at a minimum) make them private.

File: exporters/ostream/include/opentelemetry/exporters/ostream/log_exporter.h
Comment (xukaren) [reply exporters/ostream/include/opentelemetry/exporters/ostream/log_exporter.h]: Mmk yeah, sorry, they were supposed to be private - they're now moved to the `.cc` file

**Review by jsuereth — APPROVED** [review]

**Pull request merged**
